### PR TITLE
Add `PromoteEnvironment` handler to API server

### PR DESCRIPTION
**Pull request by devholic**
> Part of #263

This PR adds `PromoteEnvironment` handler to API server to let users promote environment with API.

### CODE REVIEW

File: internal/api/handler/promote_environment_v1alpha1.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+				Namespace:    req.Msg.GetProject(),
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion:         env.APIVersion,
+						Kind:               env.Kind,
+						Name:               env.Name,
+						UID:                env.UID,
+						BlockOwnerDeletion: pointer.Bool(true),
+					},
+				},
```
Comment (devholic) [inline internal/api/handler/promote_environment_v1alpha1.go:70]: Sets `OwnerReference` to cleanup promotion resources after the `environment` deleted

File: internal/api/handler/promote_environment_v1alpha1.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+			}
+		}
+		if !stateExists {
+			return nil, connect.NewError(connect.CodeNotFound,
+				fmt.Errorf("state %q not found in environment %q", req.Msg.GetState(), env.Name))
+		}
+
+		promotion := &kubev1alpha1.Promotion{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: fmt.Sprintf("%s-", req.Msg.GetName()),
```
Comment (devholic) [inline internal/api/handler/promote_environment_v1alpha1.go:60]: @krancour We discussed about naming rule in offsite meeting, but I'm still not sure if we should format promotions in same way (`<env>-to-<state_id>`). As far as I understand, the state ID should not affect the state ID hash, so I think we can use random suffix for the promotion resources 🤔 WDYT?

Comment (rbreeze) [reply internal/api/handler/promote_environment_v1alpha1.go:60]: @krancour can correct me if I'm wrong but to me it seemed that the `state_id` in the promotion resource name was more for convenience than for any technical reason. Perhaps we can truncate the state ID in the resource name instead?

**Review by rbreeze — APPROVED** [review]
LGTM but had a comment regarding naming

File: internal/api/handler/promote_environment_v1alpha1.go (hunk 2 above)
Comment (krancour) [reply internal/api/handler/promote_environment_v1alpha1.go:60]: The name of a `Promotion` resource doesn't affect its outcome in any way. It's the only the `spec.state` field that matters.

The rationale for the naming scheme, however, is that if new state is discovered for a given `Environment` and auto-promotion is enabled, a predictable naming scheme enables the `Environment` reconciler to check whether the `Promotion` it intends to create might already exists, which is a thing that could happen if a previous attempt to reconcile the `Environment` failed late in the loop.

For this reason, the controller will have to continue using that predictable naming scheme until/unless we find a better way to ensure we never automatically create duplicate `Promotion`s.

I think using `GenerateName` here in the API server makes sense.

So... LGTM.

Aside:

> As far as I understand, the state ID should not affect the state ID hash

Just to clarify this, a state ID itself _is_ a hash. The IDs of materials that make up the state are concatenated in a canonical order and then hashed.

**Comment by krancour** [comment]
LGTM, but be sure to read my clarification on state IDs and `Promotion` naming conventions just in case it raises any concerns.

**Pull request merged**